### ISSUE TITLE
reduced-round crypto_stream_salsa20*()  slower than 20-round salsa

### ISSUE THREAD

**Issue by ivmaykov** [jedisct1/libsodium#622]
TLDR: reduced-round salsa versions are slower than full-round version.
Libsodium version is 1.0.13, so if this is already fixed in trunk then ignore.

I was prototyping something and needed an efficient PRNG for generating a long-ish stream (several KB) of pseudo-random data, so I benchmarked several of the crypto_stream_*() options in libsodium. For some reason, the 12-round and 8-round versions of Salsa20 (crypto_stream_salsa208() and crypto_stream_salsa2012()) are slower than the full 20-round version (crypto_stream_salsa20()).

My program generated about 2.6KB of pseudorandom data with different crypto_stream_salsa20*() variants. There is some other stuff going on too, but the relevant part is the difference between the timings:

crypto_stream_salsa20(): 1.92us
crypto_stream_salsa2012(): 4.02us (slower by about 2.1us)
crypto_stream_salsa208(): 3.10us (slower by about 1.2us)

Maybe it's using an AVX or SSE implementation for the full round version and a pure C implementation for the reduced-round versions?

This was tested on a 24-core Intel server CPU, not sure exactly which one but here is the output of `cat /proc/cpuinfo` for the first core (the rest of the cores are identical):

```
$ cat /proc/cpuinfo
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 94
model name	: Intel Core Processor (Skylake)
stepping	: 3
microcode	: 0x1
cpu MHz		: 2394.474
cache size	: 16384 KB
physical id	: 0
siblings	: 1
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology eagerfpu pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat
bugs		:
bogomips	: 4788.94
clflush size	: 64
cache_alignment	: 64
address sizes	: 40 bits physical, 48 bits virtual
power management:
```

**Comment by jedisct1** [comment]
This is expected.

Reduced-round versions are just here for backward-compatibility with the NaCl API. The implementations are small but not optimized. They are not documented, not used by any high-level API, not compiled at all in minimal builds and their prototypes are now tagged as deprecated. They may eventually go away.

The full-round version is fast enough for most application, so the intent is to really encourage people to use it instead. While [this recent cryptanalysis](https://tosc.iacr.org/index.php/ToSC/article/viewFile/574/516) of Salsa and ChaCha doesn't lead to any practical attack, it's still a signal that having a comfortable security margin with a couple extra rounds may not be a bad idea.

**Comment by ivmaykov** [comment]
Thanks for the explanation! Closing.

**Comment by jedisct1** [comment]
That being said, you can change [this constant](https://github.com/jedisct1/libsodium/blob/4bd18252d4ba7bd088cba2bb2c656dd372dc563a/src/libsodium/crypto_stream/salsa20/xmm6int/salsa20_xmm6int-avx2.c#L29) to lower the number of rounds in the AVX2 implementation (use the `stable` branch).

Or if you really need a fast PRG on an Intel CPU, AES will be hard to beat.

**Comment by ivmaykov** [comment]
@jedisct1 for some reason crypto_stream_aes128ctr() is very very slow on my intel machine - hundreds of times slower than crypto_stream_xchacha20(). I don't know why. I am going with crypto_stream_xchacha20(), it is fast enough.

The documentation at https://download.libsodium.org/doc/advanced/salsa20.html should probably be updated, right now it says that the reduced-round versions are faster. If someone wants a really fast, reduced-security PRNG they might end up using crypto_salsa_208() on the assumption that it is faster, not realizing that it's less optimized and actually slower.

**Comment by ivmaykov** [comment]
@jedisct1 I did get AES-128-CTR from OpenSSL working and it is about 2/3rds of the speed of XChaCha20, but XChaCha is fast enough for my use case and OpenSSL is gross so I'm not going to use it :)

**Comment by jedisct1** [comment]
You probably don't need OpenSSL for that. Try something like:

```c
#include <immintrin.h>

#if defined(__INTEL_COMPILER) || defined(_MSC_VER)
# define CRYPTO_ALIGN(x) __declspec(align(x))
#else
# define CRYPTO_ALIGN(x) __attribute__ ((aligned(x)))
#endif

#define ROUNDS 10

static void
key_expand(__m128i round_keys[ROUNDS + 1], const unsigned char key[16])
{
    __m128i t, t1;
    int     i;

    i = 0;
    t = _mm_loadu_si128((const __m128i *) (const void *) key);

#define DRC(RC) do { \
    t1 = _mm_aeskeygenassist_si128(t, (RC)); \
    round_keys[i++] = t; \
    t = _mm_xor_si128(t, _mm_slli_si128(t, 4)); \
    t = _mm_xor_si128(t, _mm_slli_si128(t, 8)); \
    t = _mm_xor_si128(t, _mm_shuffle_epi32(t1, 0xff)); \
} while (0)

    DRC(1);  DRC(2);  DRC(4);   DRC(8);  DRC(16);
    DRC(32); DRC(64); DRC(128); DRC(27); DRC(54);
    round_keys[i] = t;
}

static void
stream(unsigned char *buf, size_t buf_len, __m128i round_keys[ROUNDS + 1])
{
    CRYPTO_ALIGN(16) unsigned char t[16];
    const __m128i one = _mm_set_epi64x(0, 1);
    __m128i       r;
    __m128i       s = _mm_setzero_si128();
    size_t        remaining = buf_len & ~ (size_t) 15;
    size_t        i;

#define COMPUTE_ROUNDS do { \
    r = _mm_aesenc_si128(_mm_aesenc_si128(s, round_keys[0]), round_keys[1]); \
    r = _mm_aesenc_si128(_mm_aesenc_si128(r, round_keys[2]), round_keys[3]); \
    r = _mm_aesenc_si128(_mm_aesenc_si128(r, round_keys[4]), round_keys[5]); \
    r = _mm_aesenc_si128(_mm_aesenc_si128(r, round_keys[6]), round_keys[7]); \
    r = _mm_aesenc_si128(_mm_aesenc_si128(r, round_keys[8]), round_keys[9]); \
    r = _mm_aesenclast_si128(r, round_keys[10]); \
} while (0)

    while (remaining > 0) {
        COMPUTE_ROUNDS;
        _mm_storeu_si128((__m128i *) buf, r);
        s = _mm_add_epi64(s, one);
        buf += 16;
        remaining -= 16;
    }
    remaining = buf_len & (size_t) 15;
    if (remaining > 0) {
        COMPUTE_ROUNDS;
        _mm_store_si128((__m128i *) t, r);
        for (i = 0; i < remaining; i++) {
            buf[i] = t[i];
        }
        s = _mm_add_epi64(s, one);
    }
    COMPUTE_ROUNDS;
    _mm_store_si128((__m128i *) t, r);
    key_expand(round_keys, t);
}
```

**Comment by jedisct1** [comment]
The last set of rounds replaces the key, for forward secrecy and so you can keep calling `stream()` as many times as needed. Remove them if you don't need that.

**Comment by ivmaykov** [comment]
@jedisct1 why is crypto_stream_aes128ctr() so slow? Is it a pure-software implementation of AES or something?

**Comment by ivmaykov** [comment]
Nevermind, I see that stream cipher is not even present in HEAD.

**Comment by jedisct1** [comment]
It was a pure software implementation. Probably the fastest software implementation up to date with no side channels, but still terribly slow compared to hardware-accelerated ones.

**Comment by jedisct1** [comment]
By the way, since a comment is not ideal for this, I just started this: https://github.com/jedisct1/aes-stream

Time for me to go to bed; I'll add some optimizations later.

**Comment by jedisct1** [comment]
Couldn't wait; implemented 8 and 2 blocks.
